src/Features/Core/Portable/NavigateTo/INavigateToSearcherHost.cs

```diff
@@ -0,0 +1,119 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+
+namespace Microsoft.CodeAnalysis.NavigateTo
+{
+    /// <summary>
+    /// Host interface abstracting over all the external functionality the <see cref="NavigateToSearcher"/> needs. This
+    /// provide an easy entry point for swapping out functionality of the host, including for testing purposes.
+    /// </summary>
+    internal interface INavigateToSearcherHost
+    {
+        INavigateToSearchService? GetNavigateToSearchService(Project project);
+        ValueTask<bool> IsFullyLoadedAsync(CancellationToken cancellationToken);
+    }
+
+    internal class DefaultNavigateToSearchHost : INavigateToSearcherHost
+    {
+        private readonly Solution _solution;
+        private readonly IAsynchronousOperationListener _asyncListener;
+        private readonly CancellationToken _disposalToken;
+
+        /// <summary>
+        /// Single task used to both hydrate the remote host with the initial workspace solution,
+        /// and track if that work completed.  Prior to it completing, we will try to get all
+        /// navigate-to requests from our caches.  Once it is populated though, we can attempt to
+        /// use the latest data instead.
+        /// </summary>
+        private static readonly object s_gate = new();
+        private static Task? s_remoteHostHydrateTask = null;
+
+        public DefaultNavigateToSearchHost(
+            Solution solution,
+            IAsynchronousOperationListener asyncListener,
+            CancellationToken disposalToken)
+        {
+            _solution = solution;
+            _asyncListener = asyncListener;
+            _disposalToken = disposalToken;
+        }
+
+        public INavigateToSearchService? GetNavigateToSearchService(Project project)
+            => project.GetLanguageService<INavigateToSearchService>();
+
+        public async ValueTask<bool> IsFullyLoadedAsync(CancellationToken cancellationToken)
+        {
+            var service = _solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+
+            // We consider ourselves fully loaded when both the project system has completed loaded
+            // us, and we've totally hydrated the oop side.  Until that happens, we'll attempt to
+            // return cached data from languages that support that.
+            var isProjectSystemFullyLoaded = await service.IsFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+            if (!isProjectSystemFullyLoaded)
+                return false;
+
+            var isRemoteHostFullyLoaded = GetRemoteHostHydrateTask().IsCompleted;
+            return isRemoteHostFullyLoaded;
```

Comment:
intentionally not 'await'ing the tasks.  we're not waiting for hydration to finish, we're just asking if hydration has finished.  so we only care about completion here.